### PR TITLE
v4 API: PKCE OAuth

### PR DESCRIPTION
**Pull request by n7studios**
## Summary

Adds support for the v4 API's PKCE OAuth authentication.

Further PR's to follow adding v4 API methods and reinstating tests.

## Testing

- `testGetOAuthURL`: Test that get_oauth_url() returns the correct URL to begin the OAuth process.
- `testGetAccessToken`: Test that get_access_token() returns the expected data.
- `testGetAccessTokenWithInvalidAuthCode`: Test that supplying an invalid auth code when fetching an access token returns a WP_Error.
- `testRefreshToken`: Test that refresh_token() returns the expected data.
- `testRefreshTokenWithInvalidToken`: Test that supplying an invalid refresh token when refreshing an access token returns a WP_Error.
- `testInvalidAPICredentials`: Test that supplying invalid API credentials to the API class returns a WP_Error.

## Checklist

* [x] I have [written a test](TESTING.md#writing-an-acceptance-test) and included it in this PR
* [x] I have [run all tests](TESTING.md#run-tests) and they pass
* [x] The code passes when [running the PHP CodeSniffer](TESTING.md#run-php-codesniffer)
* [x] Code meets [WordPress Coding Standards](DEVELOPMENT.md#coding-standards) for PHP, HTML, CSS and JS
* [x] [Security and Sanitization](DEVELOPMENT.md#security-and-sanitization) requirements have been followed
* [x] I have assigned a reviewer or two to review this PR (if you're not sure who to assign, we can do this step for you)

### CODE REVIEW

File: src/class-convertkit-api.php [hunk 1]
```diff
@@ -2036,66 +2290,103 @@ private function request( $endpoint, $method = 'get', $params = array(), $retry_
...
+			}
+
+			$this->log( 'API: Error: ' . $error );
+
+			switch ( $http_response_code ) {
+				// If the HTTP response code is 401, and the error matches 'The access token expired', refresh the access token now
+				// and re-attempt the request.
+				case 401:
+					if ( $error !== 'The access token expired' ) {
+						break;
```
Comment (noelherrick) [inline src/class-convertkit-api.php:2357]: Would this suppress an error?

**Review by noelherrick — APPROVED** [review]

**Comment by noelherrick** [comment]
@mercedesb Just tagging you in around the code challenge parts

File: src/class-convertkit-api.php (hunk 1 above)
Comment (n7studios) [reply src/class-convertkit-api.php:2357]: No; it'll break out of the switch statement, meaning the next code executed will return a `WP_Error` object:
https://github.com/ConvertKit/convertkit-wordpress-libraries/pull/36/files/9a5c6a0a363a2da6a77b3fb4d1ce32d4eac69bf6#diff-f92d1b708df139635b79bf42702baa5c2f612a5a742fcc2793e40661cc71a6ceR2387

**Review by mercedesb — APPROVED** [review]
The OAuth bits look good to me

**Pull request merged**